[lake] Use file to store lake snapshot

luoyuxia opened the pull request:
<!--
*Thank you very much for contributing to Fluss - we are happy that you want to help us improve Fluss. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GitHub issue](https://github.com/apache/fluss/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no issue.

  - Name the pull request in the format "[component] Title of the pull request", where *[component]* should be replaced by the name of the component being changed. Typically, this corresponds to the component label assigned to the issue (e.g., [kv], [log], [client], [flink]). Skip *[component]* if you are unsure about which is the best component.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.

  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes.

  - Each pull request should address only one issue, not mix up code from multiple issues.


**(The sections below can be removed for hotfixes or typos)**
-->

### Purpose

<!-- Linking this pull request to the issue -->
Linked issue: close #2060 

<!-- What is the purpose of the change -->

### Brief change log
- Change to store lake snapshot info to a file and record the file handle in zk node instead of the full lake snapshot
- Change commit lake snapshot to be asynchronously since it'll have heavy io

<!-- Please describe the changes made in this pull request and explain how they address the issue -->

### Tests

<!-- List UT and IT cases to verify this change -->

### API and Format

<!-- Does this change affect API or storage format -->

### Documentation

<!-- Does this change introduce a new feature -->

wuchong commented on fluss-common/src/main/java/org/apache/fluss/utils/FlussPaths.java:
```diff
@@ -681,6 +683,28 @@ public static FsPath remoteKvSnapshotDir(FsPath remoteKvTabletDir, long snapshot
+     * <p>The path contract:
+     *
+     * <pre>
+     * {$remote.data.dir}/lake/{databaseName}/{tableName}-{tableId}/snapshot/{snapshotId}
```
Better to use `{snapshotId}.snapshot` as the file name to distinguish the usage of this file, we may have more files under this directory.

wuchong reviewed (commented):
1. Since the`getLatestLakeSnapshot` RPC adds one more remote storage IO, it may blocks TabletService/CoordinatorService requests queue. I suggested to having an `ioExecutor` in TabletServer as well, and move the IO operation into the `ioExecutor` and return the future. 
2. The current serialized json data of `LakeTableSnapshot` is not compacted. We should design a new serialization format to reduce the size (avoid duplicated information in the json).
3.  I suggest it's best to consider the DV table issue together, it should involve modifying the same `laketable` znode format.

wuchong commented on fluss-server/src/main/java/org/apache/fluss/server/coordinator/CoordinatorEventProcessor.java:
```diff
@@ -1146,50 +1172,52 @@ private <T> void processAccessContext(AccessContextEvent<T> event) {
-        return response;
+                        // send notify lakehouse data request to all replicas via coordinator event
+                        coordinatorEventManager.put(
+                                new NotifyLakeTableOffsetEvent(lakeTableSnapshots));
```
I'm not sure about that. I think only update when snapshot id is larger is definitely no risk. I see current approach may risk data consistency problem when there is zombie tiering service.

luoyuxia replied:
> @luoyuxia, is the snapshot ID still a `long` value and guaranteed to be unique? Is it the same in Delta Lake?

Yes, the snapshot ID still a long value and guaranteed to be unique. Delta Lake's snapshot id(called version in delta) is strictly incremental.

wuchong replied:
How iceberg to check which snapshot is newer? Use snapshot timestamp?

luoyuxia replied:
I think snapshot timestamp should work for most case except for the corner case that multiple snapshot commit in same timestamp-ms.
Iceberg  maintains a history with snapshot IDs which helps establish order. In this order, it can compare snapshot although time-consuming.

luoyuxia commented:
@wuchong Hi, I already addressed your comment.
1. Remove unnecessary information in `LakeTableSnapshot` to reduce store size
2. Store a list of snapshot in zk node, each snapshot records the tiered offset and readable offset.  In this pr, we always set tiered offset to equal to readable offset, and only keep one snapshot since if not consider dv fix, we only need one. But the dv fix issue, I'll extend it to contains multiple  lake snapshot. 
3. For comment https://github.com/apache/fluss/pull/2037#discussion_r2580518738, maybe we can address it in another pr?

wuchong commented on fluss-server/src/main/java/org/apache/fluss/server/coordinator/CoordinatorServer.java:
```diff
@@ -407,6 +399,11 @@ CompletableFuture<Void> stopServices() {
 
+            if (ioExecutor != null) {
+                // shutdown io executor
+                ExecutorUtils.gracefulShutdown(5, TimeUnit.SECONDS, ioExecutor);
```
We should keep the original try-catch wrapper, and handle potential exceptions?

luoyuxia replied:
Miss it. I think we can follow the reasource release code pattern in here.

wuchong commented on fluss-common/src/main/java/org/apache/fluss/config/ConfigOptions.java:
```diff
@@ -394,6 +394,14 @@ public class ConfigOptions {
                                     + "for fault tolerance. Examples: `RACK1`, `cn-hangzhou-server10`");
 
+    public static final ConfigOption<Integer> TABLET_SERVER_IO_POOL_SIZE =
+            key("tablet-server.io-pool.size")
```
Can we unify the io-pool for coordinator and tablet servrer? We can rename this config to `server.io-pool.size` in this PR. And create follow-up issues:
1. deprecate `kv.snapshot.transfer-thread-num` and `remote.log.data-transfer-thread-num`, and use `server.io-pool.size` instead, may turn it into 10 threads by default. 
2. deprecate `coordinator.io-pool.size` and use `server.io-pool.size` instead.

wuchong commented on fluss-server/src/main/java/org/apache/fluss/server/coordinator/RemoteStorageCleaner.java:
```diff
@@ -57,10 +61,16 @@ public RemoteStorageCleaner(Configuration configuration, ExecutorService ioExecu
         if (isKvTable) {
             asyncDeleteDir(FlussPaths.remoteTableDir(remoteKvDir, tablePath, tableId));
         }
+        if (isLakeEnabled) {
```
If a table was enabled datalake, but turned off later, and then the table was deleted. We may leave the lake snapshot metadata file here? Maybe we can simply do the delete for the metadata file, no matter it is datalake enabled or not? (add comment for this if we do such change)

wuchong commented on fluss-common/src/main/java/org/apache/fluss/config/ConfigOptions.java:
```diff
@@ -769,13 +786,19 @@ public class ConfigOptions {
+     *     instead.
+     */
+    @Deprecated
     public static final ConfigOption<Integer> REMOTE_LOG_DATA_TRANSFER_THREAD_NUM =
```
There is still a usage in the test code. We should update the test code to use the new config.

wuchong commented on fluss-server/src/main/java/org/apache/fluss/server/zk/data/lake/LakeTable.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Nullable
+    public LakeSnapshotMetadata getLakeTableLatestSnapshot() {
```
The method name confuses a lot, because it doesn't returns a `LakeTableSnapshot`. Better to name `getLatestLakeSnapshotMetadata`.

wuchong commented on fluss-server/src/main/java/org/apache/fluss/server/zk/data/lake/LakeTable.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Nullable
+    public List<LakeSnapshotMetadata> getLakeSnapshotMetadata() {
```
It returns a list of `LakeSnapshotMetadata`, better to name `getLakeSnapshotMetadatas`.

wuchong reviewed (commented):
As we have deprecated some options with new options, could you add some statement for this in the upgrade notes?
